testing/trino-testing/src/main/java/io/trino/testing/ThreadAssertions.java

```diff
@@ -0,0 +1,98 @@
+
+import io.trino.jvm.Threads;
+
+import java.lang.management.ManagementFactory;
+import java.lang.management.ThreadMXBean;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static io.airlift.concurrent.MoreFutures.getFutureValue;
+import static java.util.stream.Collectors.joining;
+
+public final class ThreadAssertions
+{
+    private ThreadAssertions() {}
+
+    private static final AtomicInteger sequence = new AtomicInteger();
+
+    public static void assertNoThreadLeakedInPlanTester(ThrowingSupplier<PlanTester> resourceCreator, Consumer<? super PlanTester> exerciseResource)
+            throws Exception
+    {
+        assertNoThreadLeaked(resourceCreator, exerciseResource);
+    }
+
+    public static <T extends QueryRunner> void assertNoThreadLeakedInQueryRunner(ThrowingSupplier<T> resourceCreator, Consumer<? super T> exerciseResource)
+            throws Exception
+    {
+        assertNoThreadLeaked(resourceCreator, exerciseResource);
+    }
+
+    private static <T extends AutoCloseable> void assertNoThreadLeaked(
+            ThrowingSupplier<T> resourceCreator,
+            Consumer<? super T> exerciseResource)
+            throws Exception
+    {
+        // warm up all statically initialized threads
+        try (T resource = resourceCreator.get()) {
+            exerciseResource.accept(resource);
+        }
+
+        ThreadMXBean threadMXBean = ManagementFactory.getThreadMXBean();
+        CompletableFuture<Void> testFuture = new CompletableFuture<>();
+        ThreadGroup threadGroup = new ThreadGroup("test-group-" + sequence.incrementAndGet());
+        Thread.ofPlatform().group(threadGroup).start(() -> {
+            testFuture.completeAsync(() -> {
+                try (T resource = resourceCreator.get()) {
+                    exerciseResource.accept(resource);
+                }
+                catch (Exception e) {
+                    throw new RuntimeException(e);
+                }
+                Thread[] threads = new Thread[256];
+                // TODO detect leaked virtual threads -- enumerate does not return them and they live in their own thread group
+                int count = threadGroup.enumerate(threads, true);
+                String stackTraces = Arrays.stream(threads, 0, count)
+                        .filter(thread -> thread != Thread.currentThread())
+                        // Common ForkJoinPool threads are statically managed, not considered a leak
+                        .filter(thread -> !thread.getName().startsWith("ForkJoinPool.commonPool-worker-"))
+                        // OkHttp TaskRunner is statically managed (okhttp3.internal.concurrent.TaskRunner), not considered a leak
+                        .filter(thread -> !thread.getName().equals("OkHttp TaskRunner"))
```

Comment:
What does "statically managed" mean in this context? If the okhttp thread pools are not being closed, then how is that not a leak?

Reply:
This is a pool living on a static field in the library, started lazily by the library. It would be a problem if we wanted to unload the classloader, but we already know classloader unloading won't really work (due to various libraries employing this pattern), and that's why duplicating plugin classloader was removed.